### PR TITLE
Migrate lib to coroutines and remove RX dependencies

### PR DESCRIPTION
**Pull request by Babelia13**
This PR aims to update the master branch of the repo with the changes made in the [masmovil forked repo](https://github.com/masmovil/masmini-kotlin).
It contains major changes related to a library migration to Kotlin coroutines instead of RX. Copyright information in all necessary classes has also updated.

In order to avoid conflicts, a new `master-candidate` branch has been created from the point the masmovil repo started to differ from `master` branch. When merging this PR, `master-candidate` branch should replace `master`.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 
+## [2.0.0] - 2021-05-01
+### Changed
+- Changed repo ownership to [hyperdevs-team](https://github.com/hyperdevs-team). Thanks [bq](https://github.com/bq) for all the work!
+- Changed package names from `com.bq.*` to `com.hyperdevs.*`
+
+## [1.5.0] - 2021-03-15
+### Added
+- Upgrade Kotlin to 1.4.31 and Kodein to 7.4.0, apart from other Android dependencies.
+
```
Comment (Babelia13) [inline CHANGELOG.md:29]: What do you think is the best option here? To keep the versions released in the más movil repo or to include only a new 2.0.0 version?

File: README.md [hunk 2]
```diff
@@ -1,5 +1,7 @@
 # Mini
-[![Release](https://jitpack.io/v/bq/mini-kotlin.svg)](https://jitpack.io/#bq/mini-kotlin)
+[![Release](https://jitpack.io/v/masmovil/mini-kotlin.svg)](https://jitpack.io/#masmovil/mini-kotlin)
```
Comment (Babelia13) [inline README.md:2]: Change this!

File: mini-common/src/main/resources/META-INF/proguard/masmini-common.pro [hunk 3]
```diff
@@ -0,0 +1,8 @@
+-keep class mini.codegen.** { *; }
```
Comment (Babelia13) [inline mini-common/src/main/resources/META-INF/proguard/masmini-common.pro:1]: Change the name of this file.

**Comment by Babelia13** [comment]
Change the name of the following files:
- `masmini-android.pro` 
- `masmini-common.pro`
- `masmini-kodein-android.pro`.
- `masmini-kodein.pro`.

File: CHANGELOG.md [hunk 4]
```diff
@@ -109,23 +118,24 @@ state (`success` or `failure`)
...
-[1.0.6]: https://github.com/bq/mini-kotlin/compare/1.0.5...1.0.6
-[1.0.5]: https://github.com/bq/mini-kotlin/compare/1.0.4...1.0.5
-[1.0.4]: https://github.com/bq/mini-kotlin/compare/1.0.3...1.0.4
-[1.0.3]: https://github.com/bq/mini-kotlin/compare/1.0.2...1.0.3
-[1.0.2]: https://github.com/bq/mini-kotlin/compare/1.0.1...1.0.2
-[1.0.1]: https://github.com/bq/mini-kotlin/compare/1.0.0...1.0.1
-[1.0.0]: https://github.com/bq/mini-kotlin/releases/tag/1.0.0
+[Unreleased]: https://github.com/hyperdevs-team/mini-kotlin/compare/2.0.0...HEAD
+[2.0.0]: https://github.com/hyperdevs-team/mini-kotlin/compare/1.4.0...2.0.0
+[1.4.0]: https://github.com/hyperdevs-team/mini-kotlin/compare/1.3.3...1.4.0
```
Comment (estefania-sarasola-ts) [inline CHANGELOG.md:123]: If we finally maintain the 1.5.0 version, remember to add it here too

File: mini-processor/src/main/java/mini/processor/ReducersGenerator.kt [hunk 5]
```diff
@@ -1,84 +1,215 @@
...
+
+class ContainerModel(element: Element) {
+    val typeName: TypeName
+    val stateTypeName: TypeName
+    val isStatic: Boolean
+
+    init {
+        compilePrecondition(
+            check = element.kind == ElementKind.CLASS,
+            message = "Reducers must be declared inside StoreContainer classes",
```
Comment (yamidragut) [inline mini-processor/src/main/java/mini/processor/ReducersGenerator.kt:183]: I think that this should be `StateContainer`
```suggestion
            message = "Reducers must be declared inside StateContainer classes",
```

File: mini-testing/build.gradle [hunk 6]
```diff
@@ -1,11 +1,28 @@
...
 apply plugin: 'kotlin'
-apply plugin: 'java-library'
 apply from: "../jitpack.gradle"
 
 dependencies {
-    implementation project(":mini-common")
-    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
-    implementation "org.jetbrains.kotlin:kotlin-reflect:$kotlin_version"
+    api project(':mini-common')
+    api "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
```
Comment (yamidragut) [inline mini-testing/build.gradle:24]: In `mini-common/build.gradle`  and others `build.gradle` files,this has been changed to `jdk8`,  maybe we should here too?

**Review by yamidragut — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/androidsample/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+            }
+        }
+
+        vm.flow().onEach {
+            textView.text = it.toString()
+            progressBar.visibility = if (it.loading) View.VISIBLE else View.INVISIBLE
+        }.launchInLifecycleScope()
+
+        vm.flow()
+                .select { it.loading }
```
[the remark this answers is not in the document]
Comment (Babelia13) [reply app/src/main/java/com/example/androidsample/MainActivity.kt:164]: This is properly auto-indented. When I try to change it, it comes bask to that indentation.

File: README.md (hunk 2 above)
Comment (Babelia13) [reply README.md]: Solved in b6f07929daa04d17efee1757805039f0a85e84b6

File: mini-common/src/main/resources/META-INF/proguard/masmini-common.pro (hunk 3 above)
Comment (Babelia13) [reply mini-common/src/main/resources/META-INF/proguard/masmini-common.pro]: Solved in b6f07929daa04d17efee1757805039f0a85e84b6

File: mini-common/src/main/java/mini/LoggerMiddleware.kt [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+        logger(Log.DEBUG, tag, "$prelude$upCorner$actionName")
+        logger(Log.DEBUG, tag, "$prelude$action")
+
+        //Pass it down
+        val start = System.nanoTime()
+        val outAction = chain.proceed(action)
+        val processTime = (System.nanoTime() - start) / 1000000
+
+        stores.forEachIndexed { idx, store -> afterStates[idx] = store.state }
```
[the remark this answers is not in the document]
Comment (Babelia13) [reply mini-common/src/main/java/mini/LoggerMiddleware.kt]: Solved in b6f07929daa04d17efee1757805039f0a85e84b6

File: mini-common/src/main/java/mini/LoggerMiddleware.kt [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+    override suspend fun intercept(action: Any, chain: Chain): Any {
+        if (action is SilentAction) return chain.proceed(action) //Do nothing
+
+        val isSuspending = action is SuspendingAction
+        val beforeStates: Array<Any?> = Array(stores.size) { }
+        val afterStates: Array<Any?> = Array(stores.size) { }
+        val actionName = extractClassName(action.javaClass)
+
+        stores.forEachIndexed { idx, store -> beforeStates[idx] = store.state }
```
[the remark this answers is not in the document]
Comment (Babelia13) [reply mini-common/src/main/java/mini/LoggerMiddleware.kt]: Solved in b6f07929daa04d17efee1757805039f0a85e84b6

File: mini-common/src/main/java/mini/Misc.kt [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mini
+
+fun Any?.toQuotedString(): String {
+    return if (this is String) return "\"$this\""
+    else this.toString()
+}
```
[the remark this answers is not in the document]
Comment (Babelia13) [reply mini-common/src/main/java/mini/Misc.kt]: Solved in b6f07929daa04d17efee1757805039f0a85e84b6

File: mini-common/src/main/java/mini/Resource.kt [hunk 11]
```diff
@@ -124,41 +113,40 @@ inline fun <T> Resource<T>.onLoading(crossinline action: (data: T?) -> Unit): Re
...
     return this.any { it.isFailure }
 }
 
-/** Any resource is loading. */
+/** Any task is running. */
 fun <T> Iterable<Resource<T>>.anyLoading(): Boolean {
     return this.any { it.isLoading }
 }
 
+
```
[the remark this answers is not in the document]
Comment (Babelia13) [reply mini-common/src/main/java/mini/Resource.kt]: Solved in b6f07929daa04d17efee1757805039f0a85e84b6

File: mini-common/src/main/java/mini/StoreFlow.kt [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    return distinctUntilChanged().runningReduce { prev, next ->
+        if (filter(prev, next)) {
+            fn(next)
+        }
+        next
+    }
+}
+
+/**
+ * Emit a value when the value goes `from` from to `to`.
```
[the remark this answers is not in the document]
Comment (Babelia13) [reply mini-common/src/main/java/mini/StoreFlow.kt]: Solved in b6f07929daa04d17efee1757805039f0a85e84b6

File: mini-common/src/main/java/mini/StoreFlow.kt [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+ */
+fun <T> Flow<T>.onEachChange(from: T, to: T, fn: (T) -> Unit): Flow<T> {
+    return onEachChange({ prev, next -> prev == from && next == to }, fn)
+}
+
+/**
+ * Emit when the value goes `true` from to `false` (it disables).
```
[the remark this answers is not in the document]
Comment (Babelia13) [reply mini-common/src/main/java/mini/StoreFlow.kt]: Solved in b6f07929daa04d17efee1757805039f0a85e84b6

File: mini-common/src/main/java/mini/StoreFlow.kt [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+ */
+fun Flow<Boolean>.onEachDisable(fn: (Boolean) -> Unit): Flow<Boolean> {
+    return onEachChange(from = true, to = false, fn)
+}
+
+/**
+ * Emit when the value goes `false` from to `true` (it enables).
```
[the remark this answers is not in the document]
Comment (Babelia13) [reply mini-common/src/main/java/mini/StoreFlow.kt]: Solved in b6f07929daa04d17efee1757805039f0a85e84b6

File: mini-processor/src/main/java/mini/processor/ProcessorUtils.kt [hunk 15]
```diff
@@ -68,32 +93,49 @@ infix fun TypeMirror.isSameType(other: TypeMirror?): Boolean {
...
+/**
+ * Map [java.lang.Object] to [Any]
+ */
+fun TypeName.safeAnyTypeName(): TypeName {
+    return if (this is ClassName && this == ClassName("java.lang", "Object")) {
+        Any::class.asTypeName()
+    } else {
+        this
+    }
+}
```
[the remark this answers is not in the document]
Comment (Babelia13) [reply mini-processor/src/main/java/mini/processor/ProcessorUtils.kt]: Solved in b6f07929daa04d17efee1757805039f0a85e84b6

File: mini-processor/src/main/java/mini/processor/ReducersGenerator.kt (hunk 5 above)
Comment (Babelia13) [reply mini-processor/src/main/java/mini/processor/ReducersGenerator.kt]: Solved in b6f07929daa04d17efee1757805039f0a85e84b6

File: mini-testing/build.gradle (hunk 6 above)
Comment (Babelia13) [reply mini-testing/build.gradle]: Solved in b6f07929daa04d17efee1757805039f0a85e84b6

**Pull request closed without merging**